src/v/serde/parquet/metadata.cc

```diff
@@ -0,0 +1,954 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "metadata.h"
+
+#include "utils/vint.h"
+
+#include <seastar/util/variant_utils.hh>
+
+#include <climits>
+#include <utility>
+#include <variant>
+
+namespace serde::parquet {
+
+namespace {
+template<typename IntType>
+void write_be(iobuf& buf, IntType v) {
+    IntType be_form = ss::cpu_to_be(v);
+    // NOLINTNEXTLINE(*reinterpret-cast*)
+    buf.append(reinterpret_cast<const char*>(&be_form), sizeof(v));
+}
+
+template<typename IntType>
+void write_uvint(iobuf& buf, IntType v) {
+    bytes b = unsigned_vint::to_bytes(v);
+    buf.append(b.data(), b.size());
+}
+
+enum class field_type : uint8_t {
+    boolean_true = 1,
+    boolean_false = 2,
+    i8 = 3,
+    i16 = 4,
+    i32 = 5,
+    i64 = 6,
+    f64 = 7,
+    binary = 8,
+    list = 9,
+    set = 10,
+    map = 11,
+    structure = 12,
+    uuid = 13,
+};
+
+// A Struct is a sequence of zero or more fields, followed by a stop field. Each
+// field starts with a field header and is followed by the encoded field value.
+// The encoding can be summarized by the following BNF:
+//
+// struct        ::= ( field-header field-value )* stop-field
+// field-header  ::= field-type field-id
+//
+// Compact protocol field header (short form) and field value:
+// +--------+--------+...+--------+
+// |ddddtttt| field value         |
+// +--------+--------+...+--------+
+//
+// Compact protocol field header (1 to 3 bytes, long form) and field value:
+// +--------+--------+...+--------+--------+...+--------+
+// |0000tttt| field id            | field value         |
+// +--------+--------+...+--------+--------+...+--------+
+//
+// Compact protocol stop field:
+// +--------+
+// |00000000|
+// +--------+
+class struct_encoder {
+public:
+    static inline const bytes empty_struct = {0}; // NOLINT
+
+    template<typename T>
+    void write_field(int16_t field_id, field_type type, T val) {
+        auto delta = field_id - _current_field_id;
+        constexpr int16_t max_short_form_delta = 16;
+        if (delta > 0 && delta < max_short_form_delta) {
+            write_short_form_field_header(static_cast<uint8_t>(delta), type);
+        } else {
+            write_long_form_field_header(type, field_id);
+        }
+        _current_field_id = field_id;
+        if constexpr (std::is_same_v<T, iobuf>) {
+            _buf.append(std::move(val));
+        } else {
+            _buf.append(val.data(), val.size());
+        }
+    }
+
+    iobuf write_stop() && {
+        write_be<uint8_t>(_buf, 0);
```

Comment:
not that it matters, but single byte doesn't need encoding, right?

Reply:
Correct. Let me simplify - I thought I would need more types here.